### PR TITLE
[TFLite 16x8] Tanh/Sigmoid implementation is extended to the general parameter scale

### PR DESCRIPTION
**Pull request by wwwind**
The current implementation of Tanh/Sigmoid activation functions for int16 (input/output) assume that
the parameter scale is POT (power of two).

This PR provides changes to extend it to the case when the parameter scale is not POT.
The reference implementation is used in this case.

Tests + versioning are included.
It has been tested on big networks as well.

### CODE REVIEW

**Comment by wwwind** [comment]
Hi @jianlijianli, Could please review this PR ?
These changes enable Tanh/Sigmoid int16 for the general case, when the scale parameter is not POT.
Thanks!

**Comment by jdduke** [comment]
@talumbau can you help w/ review?

File: tensorflow/lite/kernels/internal/reference/integer_ops/tanh.h [hunk 1]
```diff
@@ -57,12 +57,16 @@ inline void Tanh(int32_t input_zero_point, int32_t input_range_radius,
...
+                 int32_t input_size, const int16_t* ptr_input_data,
+                 int16_t* ptr_output_data) {
   // We use the LUT for sigmoid and take into account, that
   // tanh(x) = 2*sigmoid(2*x) - 1
+
+  int32_t input_data_mul = (input_multiplier > 0) ? input_multiplier : 1;
+
   for (int i = 0; i < input_size; ++i, ptr_input_data++, ptr_output_data++) {
-    int32_t input_data = *ptr_input_data;
+    int32_t input_data = (*ptr_input_data)*input_data_mul;
```
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/integer_ops/tanh.h:69]: ` * input_data_mul` (add spaces)

File: tensorflow/lite/kernels/internal/reference/integer_ops/logistic.h [hunk 2]
```diff
@@ -58,12 +58,15 @@ inline void Logistic(int32_t input_zero_point, int32_t input_range_radius,
...
+inline void Logistic(int32_t input_multiplier, int32_t input_size,
+                     const int16_t* ptr_input_data, int16_t* ptr_output_data) {
   // We use the LUT for sigmoid and take into account, that
   // tanh(x) = 2*sigmoid(2*x) - 1
+
+  int32_t input_data_mul = (input_multiplier > 0) ? input_multiplier : 1;
+
   for (int i = 0; i < input_size; ++i, ptr_input_data++, ptr_output_data++) {
-    int32_t input_data = *ptr_input_data;
+    int32_t input_data = (*ptr_input_data)*input_data_mul;
```
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/integer_ops/logistic.h:69]: ` * input_data_mul` (add spaces)

File: tensorflow/lite/kernels/activations.cc [hunk 3]
```diff
@@ -518,19 +526,28 @@ TfLiteStatus SigmoidPrepare(TfLiteContext* context, TfLiteNode* node) {
     TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
 
     int input_scale_log2_rounded;
-    TF_LITE_ENSURE(context,
-                   CheckedLog2(input->params.scale, &input_scale_log2_rounded));
+    bool paramScalePOT =
```
Comment (jdduke) [inline tensorflow/lite/kernels/activations.cc:529]: param_scale_po2

File: tensorflow/lite/kernels/activations.cc [hunk 4]
```diff
@@ -430,21 +430,29 @@ TfLiteStatus TanhPrepare(TfLiteContext* context, TfLiteNode* node) {
...
```
Comment (jdduke) [inline tensorflow/lite/kernels/activations.cc:433]: param_scale_pot or param_scale_po2

**Comment by wwwind** [comment]
Hi @jdduke, corrected, please take a look.

**Review by jdduke — APPROVED** [review]

**Pull request merged**